keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,821 @@
+
+1. **Kubelet Configuration:** Includes relevant kubelet feature gates and configurations. A single logical capability can represent multiple underlying settings.
+2. **Runtime Configurations:** Covers scheduler-relevant runtime details like supported handlers and features, obtained via the existing [Status RPC](https://github.com/kubernetes/cri-api/blob/79a12c13b6e0a0e049d166d72b0f352057d9e41b/pkg/apis/runtime/v1/api.proto#L117) between Kubelet and CRI.
+3. **Kernel Configurations:** Details kernel support necessary for certain pod features, which may be optional or available only in newer kernel versions.
+
+**NodeCapabilities Requirements:**
+
+1. Must be derived from static configuration.
+2. Do not indicate operational health or Node readiness.[ Node Readiness Gates](https://docs.google.com/document/d/11i2_rewvcbQkFFq1BIwHa7lgIefgZ8Mak-QMZjP7fFs/edit?tab=t.0)<span style="text-decoration:underline;"> </span>are better suited for such dynamic readiness signals.
+3. Usable during scheduling or API request validation to:
+    *   Filter nodes in Kube-Scheduler by matching a pod's needs (from its spec) to Node Capabilities.
+    *   Or, Infer from API requests (Eg: in-place resize or ephemeral containers) at the validation stage. 
+
+## User Stories 
+
+### Story 1
+
+Effectively manage version skew during new feature introductions, and enable faster feedback loops with safer rollouts for Kubelet/Runtime features by scheduling Pods to nodes possessing these specific capabilities, especially in mixed-version clusters.
+
+**Examples**: 
+
+*   Alpha features like[ PodLevelResources](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pod-level-resources/) which may be enabled on selected Node Pools for evaluation. Instead of Node labels and selectors, the kube-scheduler can use` pod.spec.resources` to map pods needing the feature to Nodes advertising the capability
+*   Beta features like[ InPlacePodResizing](https://kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/) may be enabled in the Control Plane but not on all the Nodes due to version skew. There currently exists a[ workaround](https://github.com/kubernetes/kubernetes/blob/23258f104d74c6f27fd4db94940d745d9d463a8f/pkg/apis/core/validation/validation.go#L5796) to handle this version skew by looking for alternate signals from the pod spec. But having such workarounds may not always be possible. Alternatively, we could use` pod.spec.nodeName` and validate if the Node has the required capability.
+
+
+### Story 2
+
+Prevent Pods from being scheduled on Nodes lacking their required Kubelet, Kernel, or Runtime features, avoiding errors, degraded performance, or silent failures.
```

Comment:
Some features depend on both a feature gate and configuration. For example, swap depends on:
1. Feature gate enabled
2. Kubelet configuration (`memorySwapBehavior`)
3. Kernel configuration (swap controller available & cgroups v2)

Would this be exposed as several capabilities, or would we collapse all 3 to a single capability representing whether swap is supported?